Add resource support to observable source assets

smackesey opened the pull request:
### Summary & Motivation

Allows provision of resources to observable source assets via either `required_resource_keys` (old-style) or Pythonic resource args.

Threading `required_resource_keys` through both `@observable_source_asset` and `SourceAssetDefinition` was awkward. If we allow config schema for observation functions (which we should) it will only get worse. The solution used here is to make `_required_resource_keys` a private arg on `SourceAsset` to avoid committing on some API questions discussed below.

Another possible approach would be to allow source asset to take a `node_def` (matching the `AssetsDefinition` API) instead of an `observe_fn`. That would technically be a break we wouldn't normally make, since `observe_fn` is not marked as an experimental arg on `SourceAsset`-- but I'd argue it is widely understood to be experimental since `@observable_source_asset` is, which is probably what anyone using this feature is using anyway.

### How I Tested These Changes

New unit tests

smackesey commented:
Current dependencies on/for this PR:
* master
  * **PR #11993** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/11993" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/11993?utm_source=stack-comment).

sryza reviewed (commented):
In the future we might want to introduce `@loadable_source_asset`:

```python
@loadable_source_asset
def my_csv_file():
    return pd.read_csv("my_csv_file.csv")
```

It's possible that a loadable source asset would also be observable.

The reason this is relevant to this PR: would we want the same set of required resource keys for both? If not, then maybe we should name this `observation_fn_required_resource_keys` or something?

On the other hand, it's likely that the set of required resource keys would be the same.  E.g. if your source asset is an S3 file, you likely need an S3 resource both to load it and to observe it.

smackesey commented:
@sryza 

>In the future we might want to introduce @loadable_source_asset

This is yet one more reason I think we should have `SourceAsset` take a `node_def` (`observation_node_def`) instead of observation fn-- complexity is going to multiply on the `SourceAsset` signature otherwise. What do you think of this idea?

sryza commented:
> This is yet one more reason I think we should have SourceAsset take a node_def (observation_node_def?) instead of observation_fn-- complexity is going to multiply on the SourceAsset signature otherwise. What do you think of this idea?

I definitely see the advantage here, but I think the reason that I had concerns about this originally is still relevant - we might want to invoke inside a sensor function. Also, it opens up a fairly wide API surface area - e.g. you'd now be able to add a `code_version` to your `observation_fn`.

Curious if @schrockn has thoughts, as he's been thinking about this area a bit recently.

schrockn commented:
I am still in the early stages of digging more deeply into observable source assets and I was not privy to all of the discussions around their design and implementation. But with these first impressions, my general thinking is that they feel and act much more like "normal" assets, rather than be a separate concept with separate language. Suggestions of naming things like `observation_fn_required_resource_keys` are definitely red flags to me that we are complecting two ideas and complicating things unnecessarily, rather than building up via composable parts. Another example red flag is the conversation in Slack where we have to contort `AssetSelection` to deal with observable source assets and normal assets in very different ways.

We have a few notions of assets.

SDA: A definition, in code, of an asset that _should_ exist. This traditional SDA definition also incorporates some very opinionated semantics in the I/O manager system.
Observable Source Asset: A definition, in code, of an asset that we _believe_ exists, but want to do computation verify to that it exists and validate/assert claims about it.
Source Asset: A definition in code of an asset that I claim exists.
Loadable Source Asset (Hypothesized above) A special sort of observable asset that automatically makes a bunch of claims given an in-memory representation of that asset returned to Dagster.

I don't think observable source assets are distinctly different enough from normal SDAs to warrant a different concept. I think they are much more similar to traditional assets rather than source assets, actually, despite their name and my definitions above. 

Instead, I think we should slightly loosen the constraints on the traditional SDAs to allow for this use case. Namely, you should be able to write an SDA where the op that SDA refers to _only_ emits an AssetObservation with attached logical version data. It, by convention, would return `Nothing` and have no associated I/O manager. (Note: I believe that currently we do emit an AssetMaterialization for Nothing outputs, but I don't think that makes sense and we should change that)

This is also connected to some thinking and exploration I've been doing to support external execution environments where the "contract" between the Dagster orchestrator and transformation is much more about the event stream, rather than our in-memory Python constructs. The top-of-mind use cases for me right now are Spark jobs and external K8s pods in different docker images. We we should be able to launch into external environments that in turn do the actual compute and read/write to storage, and then either (a) have those environments write to event log directly or (b) have the orchestrating process to do it on their behalf in an easy-to-use, well-supported way.

What connects these two use cases––source assets and external compute environments––is that it is no longer the I/O manager that is controlling writing asset materializations and observations. Instead compute is just emitting those events and taking over responsibility for I/O.

In this way of thinking I/O managers are very and ergonomic convenient ways to build a graph of assets in a single environment that abide by the event stream rules, but are not required.

sryza commented:
@schrockn do you have thoughts about the specific implications of your point for our Python APIs and/or Dagit?

I agree with you that there's a lot of thematic resemblance between these categories and opportunity to make the experience more consistent.

That said, I think there's a fundamental difference between:
1. an asset whose existence we know about at definition time
2. an asset whose existence we know about at definition time AND that we know how to materialize

The latter type supports specific, important operations in Python and in Dagit that the former type does not, and consequently I think that overly blurring the distinction between the two risks degrading the user experience.

sryza commented:
To add some specificity, I think it's very important that, before taking an action to "run an asset", a user can tell whether doing so is going to materialize it or just observe it.

sryza commented:
> @schrockn do you have thoughts about the specific implications of your point for our Python APIs and/or Dagit?

@schrockn sorry I realized I didn't read some of your points in enough detail and that you did include specific implications.

> Note: I believe that currently we do emit an AssetMaterialization for Nothing outputs, but I don't think that makes sense and we should change that

IMO the way this works right now makes a lot of sense.  I think that making this change would be a strong hit to user ergonomics.  It would effectively require them to redundantly specify their asset key in situations where they currently only need to specify them once, which adds boilerplate and opportunity for error.

schrockn commented:
> IMO the way this works right now makes a lot of sense. I think that making this change would be a strong hit to user e ergonomics. It would effectively require them to redundantly specify their asset key in situations where they currently only need to specify them once, which adds boilerplate and opportunity for error.

Note I was suggesting that _only_ for `Nothing`, not in the general case. I don't want to sidetrack the convo here. I agree that when you are fully in I/O manager land, automatically emitting AssetMaterializations with the right asset key makes sense for all the reasons you mention.

sryza commented:
> Note I was suggesting that only for Nothing, not in the general case. I don't want to sidetrack the convo here. I agree that when you are fully in I/O manager land, automatically emitting AssetMaterializations with the right asset key makes sense for all the reasons you mention.

I understand and that's what I am explicitly arguing against

schrockn commented:
> For example, if A -> B, and I'm launching a run that observes both A and B, there's no reason to wait until A has been observed before observing B. However, if I'm launching a run that materializes both A and B, then I should wait until A has materialized to materialize B.

I'm a bit confused by this example as it implies that B has an upstream dependency on A, and observable source assets don't have upstream dependencies. I suppose that would apply if we optionally added observation functions to all assets.

Re: "important operations in Python and in Dagit" do you have other specific ones that you think are of paramount importance?

sryza commented:
> I'm a bit confused by this example as it implies that B has an upstream dependency on A, and observable source assets don't have upstream dependencies. I suppose that would apply if we optionally added observation functions to all assets.

Ah right, yes, exactly.

> Re: "important operations in Python and in Dagit" do you have other specific ones that you think are of paramount importance?

The first one that comes to mind is clicking the Materialize button. I think it would be a loss if the button had to say something generic like "run" and then you don't get to find out whether the asset has been materialized or observed until the step completes.

sryza commented:
Here's what our concept page currently says:

> An asset is an object in persistent storage, such as a table, file, or persisted machine learning model. A software-defined asset is a Dagster object that couples an asset to the function and upstream assets that are used to produce its contents.

I would be open to changing this to something like "materializable software-defined asset" if we think it makes the ontology more clear.

smackesey commented:
@sryza Thoughts on naming the `required_resource_keys` arg here `_required_resource_keys`, so that we can get this feature in before resolving the broader questions of source asset ontology?

sryza commented:
Suppose we merged the PR as-is and we later wanted to allow source assets to be loadable - would that require a breaking change?

schrockn reviewed (dismissed):
What is the pressing need for this right now?

When it comes to this observables source asset duality issue, I'm certainly in "contain the damage" mode until we have more clarity here. Supporting this, for example, adds another decorator that we have know how to inject resources into, increasing the surface area of resources v2, just an example, and adds another concept to the list of concepts that we have to bind resources to.

I'd prefer we step back before proceeding here.

sryza commented:
I don't feel particular urgency on landing this, but FWIW I feel quite confident that the future will contain both:
- SourceAssets with a built-in observation function
- The ability to supply resources to that function

smackesey commented:
>Suppose we merged the PR as-is and we later wanted to allow source assets to be loadable - would that require a breaking change?

It depends on how we implemented loadable source assets, but if we merged it as is there would be `required_resource_keys` exposed directly on `SourceAsset`. If we wanted to have separate loadable and observable ops with separate `required_resource_keys`, then we'd end up with either a breaking change or a confusing API (where `required_resource_keys` applies only to observation). That's why I support making it `_required_resource_keys` on `SourceAsset`-- so that we have flexibility to change it later.

>What is the pressing need for this right now?

Just that people expect resources to work with source asset observation right now, and they don't. I see this more as fixing a bug than adding a feature.

smackesey commented:
> Which "people"? What is their exact use case?

I've had a support request for it at least once. I assume the use case is to query a database to compute a `LogicalVersion`.

smackesey commented:
>We should make sure to support new-style resource passing to this as well.

OK, I modified this and added a test to make sure it handles pythonic resources.

sryza commented on python_modules/dagster/dagster_tests/asset_defs_tests/test_source_asset_observation_job.py:
```diff
@@ -105,3 +108,97 @@ def bar(context):
+
+
+@pytest.mark.parametrize(
+    "is_valid,resource_defs",
```
The fact that most of the logic in the function depends on this boolean makes me think that it would be clearer to split it into two separate tests, and just define the asset at module level. Not critical feedback though.

smackesey replied:
can't do that because of the closure over `executed`

sryza replied:
ah right - could have a shared function though. or it's a small enough amount of code it could just be copy/pasted. the `is_valid=False` one doesn't even need an implementation

sryza commented on python_modules/dagster/dagster/_core/definitions/source_asset.py:
```diff
@@ -74,6 +85,7 @@ class SourceAsset(ResourceAddable):
             the asset when it's used as an input to other assets inside a job.
         io_manager_def (Optional[IOManagerDefinition]): (Experimental) The definition of the IOManager that will be used to load the contents of
             the asset when it's used as an input to other assets inside a job.
+        required_resource_keys (Optional[Set[str]]): Set of resource handles required by the observe op.
```
If this is private, we shouldn't document it publicly, right?

sryza commented on python_modules/dagster/dagster/_core/definitions/source_asset.py:
```diff
@@ -59,9 +71,8 @@ def __call__(self) -> DataVersion:
-SourceAssetObserveFunction: TypeAlias = Union[
-    SourceAssetObserveFunctionWithContext, SourceAssetObserveFunctionNoContext
-]
+# Going with this catch-all for the time-being to permit pythonic resources
```
We don't need the protocols above anymore in this case, right?

sryza commented on python_modules/dagster/dagster/_core/definitions/source_asset.py:
```diff
@@ -104,6 +116,7 @@ def __init__(
         group_name: Optional[str] = None,
         resource_defs: Optional[Mapping[str, ResourceDefinition]] = None,
         observe_fn: Optional[SourceAssetObserveFunction] = None,
+        _required_resource_keys: Optional[AbstractSet[str]] = None,
```
I think worth mentioning why this has an underscore - i.e. that it's used for the observation function, but we're not sure what we want to do when an asset is both observable and can also be subject to other operations.

sryza commented on python_modules/dagster/dagster/_core/definitions/source_asset.py:
```diff
@@ -192,6 +217,11 @@ def fn(context: OpExecutionContext):
 
+    @public
+    @property
+    def required_resource_keys(self) -> AbstractSet[str]:
```
I thought we said this would be private?
